scheduler/data-flow/src/test/kotlin/io/seldon/dataflow/kafka/PipelineStatusTest.kt

```diff
@@ -0,0 +1,172 @@
+                ),
+                arguments(
+                    "StreamStopped(prevState=nested StreamStopped with Error)",
+                    !IS_ACTIVE,
+                    IS_ERROR,
+                    PipelineStatus.StreamStopped(
+                        PipelineStatus.StreamStopped(
+                            PipelineStatus.Error(KafkaStreams.State.ERROR),
+                        ),
+                    ),
+                ),
+                arguments(
+                    "Error(errorState=null)",
+                    !IS_ACTIVE,
+                    IS_ERROR,
+                    PipelineStatus.Error(null),
+                ),
+                arguments(
+                    "Error(errorState=non-error,active state)",
+                    IS_ACTIVE,
+                    IS_ERROR,
+                    PipelineStatus.Error(KafkaStreams.State.RUNNING),
+                ),
+                arguments(
+                    "Error(errorState=non-error,non-active state)",
+                    !IS_ACTIVE,
+                    IS_ERROR,
+                    PipelineStatus.Error(KafkaStreams.State.NOT_RUNNING),
+                ),
+                arguments(
+                    "Error(errorState=error state)",
+                    !IS_ACTIVE,
+                    IS_ERROR,
+                    PipelineStatus.Error(KafkaStreams.State.PENDING_ERROR),
+                ),
+                arguments(
+                    "PipelineStatus(state=error state, hasError=false)",
+                    !IS_ACTIVE,
+                    IS_ERROR,
+                    PipelineStatus(KafkaStreams.State.PENDING_ERROR, false),
+                ),
+                arguments(
+                    "PipelineStatus(state=error state, hasError=true)",
+                    !IS_ACTIVE,
+                    IS_ERROR,
+                    PipelineStatus(KafkaStreams.State.PENDING_ERROR, true),
+                ),
+                arguments(
+                    "PipelineStatus(state=non-error,non-active state, hasError=false)",
+                    !IS_ACTIVE,
+                    !IS_ERROR,
+                    PipelineStatus(KafkaStreams.State.NOT_RUNNING, false),
+                ),
+                arguments(
+                    "PipelineStatus(state=non-error,non-active state, hasError=true)",
+                    !IS_ACTIVE,
+                    IS_ERROR,
+                    PipelineStatus(KafkaStreams.State.NOT_RUNNING, true),
+                ),
+                arguments(
+                    "PipelineStatus(state=non-error,active state, hasError=false)",
+                    IS_ACTIVE,
+                    !IS_ERROR,
+                    PipelineStatus(KafkaStreams.State.CREATED, false),
+                ),
+                arguments(
+                    "PipelineStatus(state=non-error,active state, hasError=true)",
+                    IS_ACTIVE,
+                    IS_ERROR,
+                    PipelineStatus(KafkaStreams.State.CREATED, true),
```

Comment:
how can it be created and has error?

Reply:
We don't typically encounter this, but the goal of "hasError" is to mark that there has been some exception/error message recorded together with the state. It could be that it's not a fatal error (the pipeline still got created), but there was an exception and we recovered.